File: ios/Earwyrm/Core/Supabase/NotificationManager.swift
import Foundation
import Supabase

@Observable
@MainActor
final class NotificationManager {
    var notifications: [AppNotification] = []
    var unreadCount: Int = 0
    var isLoading = false
    var hasMore = true

    private let pageSize = 20

    private static let columns = """
        id, user_id, type, lyric_id, comment_id, collection_id, actor_id, \
        lyric_snippet, song_title, artist_name, share_token, actor_username, \
        comment_content, collection_name, follow_type, follow_value, \
        read_at, created_at
        """

    // MARK: - Fetch Notifications

    func fetchNotifications(userId: UUID, offset: Int = 0) async {
        isLoading = true
        do {
            let result: [AppNotification] = try await supabase
                .from("notifications")
                .select(Self.columns)
                .eq("user_id", value: userId.uuidString)
                .order("created_at", ascending: false)
                .range(from: offset, to: offset + pageSize - 1)
                .execute()
                .value

            if offset == 0 {
                notifications = result
            } else {
                notifications.append(contentsOf: result)
            }
            hasMore = result.count >= pageSize
        } catch {
            print("Fetch notifications error: \(error)")
        }
        isLoading = false
    }

    // MARK: - Unread Count

    private struct IdOnly: Decodable {
        let id: UUID
    }

    func fetchUnreadCount(userId: UUID, lastSeenAt: Date?) async {
        do {
            var query = supabase
                .from("notifications")
                .select("id")
                .eq("user_id", value: userId.uuidString)

            if let lastSeen = lastSeenAt {
                let formatter = ISO8601DateFormatter()
                formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
                query = query.gt("created_at", value: formatter.string(from: lastSeen))
            }

            let result: [IdOnly] = try await query
                .execute()
                .value

            unreadCount = result.count
        } catch {
            print("Fetch unread count error: \(error)")
        }
    }

    // MARK: - Mark All Seen

    private struct LastSeenUpdate: Encodable {
        let lastActivitySeenAt: Date

        enum CodingKeys: String, CodingKey {
            case lastActivitySeenAt = "last_activity_seen_at"
        }
    }

    func markAllSeen(userId: UUID) async {
        let now = Date()
        do {
            let update = LastSeenUpdate(lastActivitySeenAt: now)
            try await supabase
                .from("profiles")
                .update(update)
                .eq("id", value: userId.uuidString)
                .execute()

            unreadCount = 0
        } catch {
            print("Mark all seen error: \(error)")
        }
    }

    // MARK: - Send Share Notification

    func sendShareNotification(
        lyricOwnerId: UUID,
        actorId: UUID,
        actorUsername: String,
        lyric: Lyric
    ) async {
        // Don't notify yourself
        guard lyricOwnerId != actorId else { return }

        do {
            let insert = ShareNotificationInsert(
                userId: lyricOwnerId,
                type: "share",
                lyricId: lyric.id,
                actorId: actorId,
                lyricSnippet: lyric.content.count > 80
                    ? String(lyric.content.prefix(80))
                    : lyric.content,
                songTitle: lyric.songTitle,
                artistName: lyric.artistName,
                shareToken: lyric.shareToken,
                actorUsername: actorUsername
            )
            try await supabase
                .from("notifications")
                .insert(insert)
                .execute()
        } catch {
            print("Send share notification error: \(error)")
        }
    }
}

File: android/app/src/main/java/com/earwyrm/app/core/supabase/NotificationManager.kt
package com.earwyrm.app.core.supabase

import com.earwyrm.app.core.model.AppNotification
import com.earwyrm.app.core.model.Lyric
import com.earwyrm.app.core.model.ShareNotificationInsert
import io.github.jan.supabase.SupabaseClient
import io.github.jan.supabase.postgrest.postgrest
import io.github.jan.supabase.postgrest.query.Order
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.datetime.Clock
import kotlinx.datetime.Instant
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class NotificationManager @Inject constructor(private val supabase: SupabaseClient) {
    private val _notifications = MutableStateFlow<List<AppNotification>>(emptyList())
    val notifications: StateFlow<List<AppNotification>> = _notifications.asStateFlow()
    private val _unreadCount = MutableStateFlow(0)
    val unreadCount: StateFlow<Int> = _unreadCount.asStateFlow()
    private var lastSeenAt: Instant? = null

    suspend fun fetchNotifications(userId: String, limit: Int = 20) {
        try { _notifications.value = supabase.postgrest.from("notifications").select { filter { eq("user_id", userId) }; order("created_at", Order.DESCENDING); limit(limit.toLong()) }.decodeList(); updateUnreadCount() } catch (_: Exception) { }
    }
    suspend fun markAllSeen(userId: String) {
        try { val now = Clock.System.now(); supabase.postgrest.from("profiles").update({ set("last_activity_seen_at", now.toString()) }) { filter { eq("id", userId) } }; lastSeenAt = now; _unreadCount.value = 0 } catch (_: Exception) { }
    }
    fun setLastSeenAt(instant: Instant?) { lastSeenAt = instant; updateUnreadCount() }
    private fun updateUnreadCount() { val s = lastSeenAt; _unreadCount.value = if (s == null) _notifications.value.size else _notifications.value.count { it.createdAt > s } }

    suspend fun sendShareNotification(lyricOwnerId: String, actorId: String, actorUsername: String, lyric: Lyric) {
        if (lyricOwnerId == actorId) return
        try {
            supabase.postgrest.from("notifications").insert(
                ShareNotificationInsert(
                    userId = lyricOwnerId,
                    lyricId = lyric.id,
                    actorId = actorId,
                    lyricSnippet = lyric.content.take(80),
                    songTitle = lyric.songTitle,
                    artistName = lyric.artistName,
                    shareToken = lyric.shareToken,
                    actorUsername = actorUsername
                )
            )
        } catch (_: Exception) { }
    }
}
